postreise/analyze/helpers.py

```diff
@@ -279,3 +280,56 @@ def summarize_plant_to_location(df, grid):
     location_data = df.T.groupby(locations_in_df).sum().T
 
     return location_data
+
+
+def get_plant_id_for_resources_in_area(scenario, area, resources, area_type=None):
+    """Get the list of plant ids of certain resources in the specific area of a
+    scenario.
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of: *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str/list resources: one or a list of resources
+    :param str area_type: one of: *'loadzone'*, *'state'*,
+        *'state_abbr'*, *'interconnect'*
+    :return: (*list*) -- list of plant id
+    """
+    if isinstance(resources, str):
+        resource_set = set([resources])
+    else:
+        resource_set = set(resources)
+    grid = scenario.state.get_grid()
+    loadzone_set = area_to_loadzone(grid, area, area_type=area_type)
+    plant_id = grid.plant[
+        (grid.plant["zone_name"].isin(loadzone_set))
+        & (grid.plant["type"].isin(resource_set))
+    ].index.tolist()
+
+    return plant_id
+
+
+def get_storage_id_in_area(scenario, area, area_type=None):
+    """Get the list of storage ids in the specific area of a scenario
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of: *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str area_type: one of: *'loadzone'*, *'state'*,
+        *'state_abbr'*, *'interconnect'*
+    :return: (*list*) -- list of storage id
+    """
+    grid = scenario.state.get_grid()
+    loadzone_set = area_to_loadzone(grid, area, area_type=area_type)
+    loadzone_id_set = {grid.zone2id[lz] for lz in loadzone_set if lz in grid.zone2id}
+
+    storage_id = (
+        grid.storage["gen"]
+        .loc[
+            grid.storage["gen"]["bus_id"]
+            .apply(lambda x: grid.bus.loc[x, "zone_id"])
+            .isin(loadzone_id_set)
+        ]
+        .index.tolist()
+    )
```

Comment:
```python
zone_buses = grid.bus.loc[grid.bus.zone_id.isin(loadzone_id_set)].index
storage_gen = grid.storage["gen"]
storage_id = storage_gen.loc[storage_gen.bus_id.isin(zone_buses)].index.tolist()
```

Reply:
Either way, less lines vs less memory usage.

Reply:
I usually prefer less variables

Reply:
What about something like:
```python
gen = grid.storage["gen"]
storage_id = gen.loc[
	gen["bus_id"].apply(lambda x: grid.bus.loc[x, "zone_id"]).isin(loadzone_id_set)
].index.tolist()
```
?

Getting everything into one expression is nice, but parsing this nine-line expression makes my head hurt a bit. Creating a `gen` variable like this shouldn't have a memory hit since we're just pointing to the same object.

Reply:
> What about something like:
> 
> ```python
> gen = grid.storage["gen"]
> storage_id = gen.loc[
> 	gen["bus_id"].apply(lambda x: grid.bus.loc[x, "zone_id"]).isin(loadzone_id_set)
> ].index.tolist()
> ```
> 
> ?
> 
> Getting everything into one expression is nice, but parsing this nine-line expression makes my head hurt a bit. Creating a `gen` variable like this shouldn't have a memory hit since we're just pointing to the same object.

Okay, I will proceed in this way then.

Reply:
Done.